### PR TITLE
Fix task snooze and dismiss functions

### LINKED ISSUE #7726: Task list items not snoozeable
### Describe the bug
Setting a task item to `is_snoozeable` does not show the option to snooze the task.

### To Reproduce

1. Add `is_snoozeable` to a server-side task
2. Load the task list on the frontend
3. Note that no drop down option is given

### PR DESCRIPTION
**Pull request by joshuatf**
Fixes #7726

Fixes the task snooze and dismissal functions and endpoints

### Screenshots

Screen Shot 2021-09-28 at 1 36 26 PM
### Detailed test instructions:

1. Add `is_snoozeable` and `is_dismissable` to a task under `src/Features/OnboardingTasks/Tasks/`
2. View the (incomplete) task in the task list.
3. Attempt to dismiss and snooze the task
4. Make sure the task is removed from the list on snooze/dismiss
5. Attempt to undo task dismissal and make sure the task is shown again
6. Attempt to undo task snooze and make sure the task is shown again

Note that you can delete the options `woocommerce_task_list_remind_me_later_tasks` or `woocommerce_task_list_dismissed_tasks` to reset items back to their original state.

### CODE REVIEW

File: src/Features/OnboardingTasks/TaskLists.php [hunk 1]
```diff
@@ -162,16 +162,16 @@ public static function get_task( $id, $task_list_id = null ) {
 			return null;
```
Comment (joelclimbsthings) [inline src/Features/OnboardingTasks/TaskLists.php:162]: The issue I mentioned in https://github.com/woocommerce/woocommerce-admin/pull/7727 is a couple lines up, with `self::get_task_list_by_id`, which doesn't exist in the class.

File: src/API/OnboardingTasks.php [hunk 2]
```diff
@@ -833,19 +808,13 @@ public function undo_dismiss_task( $request ) {
...
 
-		$snooze_task = TaskLists::get_task( $task_id, $task_list_id );
+		$task = TaskLists::get_task( $task_id, $task_list_id );
 
-		if ( $snooze_task && isset( $snooze_task['isSnoozeable'] ) && $snooze_task['isSnoozeable'] ) {
-			$is_snoozeable = true;
-		}
-
-		if ( ! $is_snoozeable ) {
+		if ( ! $task || ! $task->is_snoozeable ) {
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:817]: May be subjective, but I mildly prefer writing this condition this way:

```suggestion
		if ( ! ( $task && $task->is_snoozeable )) {
```

Just a little clearer to me, but perhaps just me 🤷🏻‍♂️

File: src/API/OnboardingTasks.php [hunk 3]
```diff
@@ -878,29 +835,10 @@ public function snooze_task( $request ) {
...
-
-		$update = update_option( 'woocommerce_task_list_remind_me_later_tasks', $snoozed );
-
-		if ( $update ) {
-			wc_admin_record_tracks_event( 'tasklist_undo_remindmelater_task', array( 'task_name' => $id ) );
-		}
-
-		return rest_ensure_response( TaskLists::get_task( $id ) );
+		$id   = $request->get_param( 'id' );
+		$task = TaskLists::get_task( $id );
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:839]: I think you'll still want to handle the case that the task ID is invalid, and this returns null. Right now it throws an unhandled error.

**Review by joelclimbsthings — CHANGES_REQUESTED** [review]
Looks great (and slightly familiar 😆) overall @joshuatf , with a couple minor comments, and one potential error.

File: src/API/OnboardingTasks.php [hunk 4]
```diff
@@ -810,19 +794,10 @@ public function dismiss_task( $request ) {
...
-		$dismissed = get_option( 'woocommerce_task_list_dismissed_tasks', array() );
-		$dismissed = array_diff( $dismissed, array( $id ) );
-		$update    = update_option( 'woocommerce_task_list_dismissed_tasks', $dismissed );
-
-		if ( $update ) {
-			wc_admin_record_tracks_event( 'tasklist_undo_dismiss_task', array( 'task_name' => $id ) );
-		}
-
+		$id   = $request->get_param( 'id' );
 		$task = TaskLists::get_task( $id );
```
Comment (joelclimbsthings) [inline src/API/OnboardingTasks.php:798]: Also here, for handling the case of an invalid task ID.

File: src/API/OnboardingTasks.php (hunk 3 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:850]: Good call, though heads up that @louwie17 will probably be temporarily overriding this logic for backwards compatibility.

File: src/API/OnboardingTasks.php (hunk 4 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:798]: Also handled, but same comment as above.

File: src/API/OnboardingTasks.php (hunk 2 above)
Comment (joshuatf) [reply src/API/OnboardingTasks.php:828]: I think the convention used in core and WCA is typically this pattern and usually favors less parentheses so I'd recommend keeping the existing style here.

File: src/Features/OnboardingTasks/TaskLists.php (hunk 1 above)
Comment (joshuatf) [reply src/Features/OnboardingTasks/TaskLists.php:162]: Thanks!  Updated.

**Review by joelclimbsthings — APPROVED** [review]
Awesome @joshuatf , thanks for the changes! 🚢

**Pull request merged**